### PR TITLE
LPD-17026 Apply editor config contributor client extensions to alloy editor configurations

### PR DESCRIPTION
**Pull request by markocikos**
### References

- Parent story: [LPD-6597 Apply editor config contributor client extensions to alloy editor configurations](https://issues.liferay.com/browse/LPD-6597)

### What is the goal of this PR?

We are implementing support for editor config contributor client extension in Alloy Editor. We are extending CX workspace sample to cover AE config format and applying it to Blogs content.

What is notably **not** in this PR are functional tests. This is blocked until automated CX deployment is available. [Related PR](https://github.com/liferay-frontend/liferay-portal/pull/3959).

### What does it look like?

image

### CODE REVIEW

File: workspaces/liferay-sample-workspace/client-extensions/liferay-sample-editor-config-contributor/client-extension.yaml [hunk 1]
```diff
@@ -3,6 +3,7 @@ assemble:
       into: static
 liferay-sample-editor-config-contributor:
     editorConfigKeys:
+        - contentEditor
```
Comment (markocikos) [inline workspaces/liferay-sample-workspace/client-extensions/liferay-sample-editor-config-contributor/client-extension.yaml:6]: I don't know any Alloy Editor usage under FI responsibility, so I chose Blogs, my typical go-to for checking AE. Applying sample to Blogs or some similar location is still probably better than building a new AE sample module.

File: workspaces/liferay-sample-workspace/client-extensions/liferay-sample-editor-config-contributor/src/index.ts [hunk 2]
```diff
@@ -9,6 +9,33 @@ import {
...
+			buttons: Array<string>;
+			name: string;
+		}
+
+		const textSelection: ISelection = toolbars.styles?.selections?.find(
+			(selection: ISelection) => selection.name === 'text'
+		);
+
+		if (textSelection.buttons) {
+			textSelection.buttons.push('video');
```
Comment (markocikos) [inline workspaces/liferay-sample-workspace/client-extensions/liferay-sample-editor-config-contributor/src/index.ts:28]: `AI Creator` plugin does not work with AE. This is unrelated to CXs and the scope of this task, so I chose a simpler existing button to add.

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
ci:stop

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/utils/client_extensions/loadEditorClientExtensions.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+import loadClientExtensions from './loadClientExtensions';
+
+interface IConfig {
+	editorTransformerURLs: Array<string>;
+}
+
+export default function loadEditorClientExtensions({
```
Comment (markocikos) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/utils/client_extensions/loadEditorClientExtensions.ts:12]: @dsanz We talked about this in https://github.com/liferay-frontend/liferay-portal/pull/3902#discussion_r1435298104. I decided to put the util in `frontend-js-web`, for simplicity and consistency. It's probably not wise to fragment places that fill `Liferay.Util` namespace.

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
ci:report:2536277819

**Comment by markocikos** [comment]
No unique failures, based on https://github.com/liferay-frontend/liferay-portal/pull/3979#issuecomment-1933825658. Waiting now for peer review.

File: workspaces/liferay-sample-workspace/client-extensions/liferay-sample-editor-config-contributor/src/index.ts (hunk 2 above)
Comment (dsanz) [reply workspaces/liferay-sample-workspace/client-extensions/liferay-sample-editor-config-contributor/src/index.ts:28]: I think it'd be clearer if we separate the samples, one for AE and the other for CKEditor. This way, each would make its own assumptions about the internal structure of the `config` object.

However, we don't want to promote AE usage, that is, adding a separate sample for AE looks "too much" advertisement. 

From the  PoV of the implementor, making this distinction in the same sample may be useful in some use cases (e.g. add the same button throughout the entire set of DXP editors).

It's a little bit more cumbersome to justify adding different buttons to different editors, in the same sample. The "video" button is however already present in the base toolbar for ckeditor so we can't change the sample to add it to CKEditor :(

All that said, we can reorganize this sample later on if we find a better way, we can move on with this.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/utils/client_extensions/loadEditorClientExtensions.ts (hunk 3 above)
Comment (dsanz) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/utils/client_extensions/loadEditorClientExtensions.ts:12]: nice!

Looking at it, I wonder whether we shall have "spetialized" versions of `load*ClientExtensions()` in the front end, besides the generic one.

The editor CX case deserves this for sure as we have a very specific, reusable semantics across different editor flavors, i.e. multiple injection points for the same CX type (AE, CKEditor), with the additional peculiarity that more than one CX instance may be loaded for the same injection point at runtime.

Normally, the situation is the opposite. We have only one JS-based injection point per CX type (which injects 0 or 1 CX instance each time), so it's general-purpose form `loadClientExtension()` is enough. As soon as we need the same injection logic in separate places, we'll need to build the corresponding abstraction and make it available.

File: workspaces/liferay-sample-workspace/client-extensions/liferay-sample-editor-config-contributor/client-extension.yaml (hunk 1 above)
Comment (dsanz) [reply workspaces/liferay-sample-workspace/client-extensions/liferay-sample-editor-config-contributor/client-extension.yaml:6]: I'm fine tiwh this. I don't think we have to invest in a AE sample in 2024

File: workspaces/liferay-sample-workspace/client-extensions/liferay-sample-editor-config-contributor/src/index.ts (hunk 2 above)
Comment (markocikos) [reply workspaces/liferay-sample-workspace/client-extensions/liferay-sample-editor-config-contributor/src/index.ts:28]: > From the PoV of the implementor, making this distinction in the same sample may be useful in some use cases (e.g. add the same button throughout the entire set of DXP editors).

Yeah, that was my line of thinking: this highlights the fact that the same CX can be applied to both Alloy Editor and CKEditors. You need to jump through few hoops, i.e. deal with different config formats, but it's possible. But, I see your point about simplicity, we are trading off simplicity to achieve this.

**Comment by markocikos** [comment]
ci:forward:force
